### PR TITLE
Specify designated vendor in sales order line

### PR DESCRIPTION
**Pull request by av-ps**
refs: https://github.com/metasfresh/metasfresh/issues/11797

### CODE REVIEW

File: backend/de.metas.dlm/base/src/main/sql/postgresql/system/5606530_sys_gh11797_C_OrderLine_C_BPartner_Vendor_ID.sql [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
```
Comment (metas-ts) [inline backend/de.metas.dlm/base/src/main/sql/postgresql/system/5606530_sys_gh11797_C_OrderLine_C_BPartner_Vendor_ID.sql:72]:
```suggestion
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/sql/postgresql/system/48-de.metas.order/5606540_sys_gh11797_C_Order_CreatePOFromSOs_IsVendorInOrderLinesRequired_param.sql [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/sql/postgresql/system/48-de.metas.order/5606540_sys_gh11797_C_Order_CreatePOFromSOs_IsVendorInOrderLinesRequired_param.sql:22]:
```suggestion
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/sql/postgresql/system/48-de.metas.order/5606540_sys_gh11797_C_Order_CreatePOFromSOs_IsVendorInOrderLinesRequired_param.sql [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/sql/postgresql/system/48-de.metas.order/5606540_sys_gh11797_C_Order_CreatePOFromSOs_IsVendorInOrderLinesRequired_param.sql:94]:
```suggestion
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/sql/postgresql/system/48-de.metas.order/5606670_sys_gh11797_AD_Message_skippedOrderLines.sql [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/sql/postgresql/system/48-de.metas.order/5606670_sys_gh11797_AD_Message_skippedOrderLines.sql:22]:
```suggestion
```

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/process/C_Order_CreatePOFromSOs.java [hunk 5]
```diff
@@ -123,6 +129,17 @@ protected String doIt() throws Exception
 		}
 		workpackageAggregator.closeAllGroups();
 
+		final Set<OrderLineId> skippedSalesOrderLines = workpackageAggregator.getSkippedSalesOrderLineIds();
+
+		if (skippedSalesOrderLines.size() > 0)
+		{
+			final UserNotificationRequest userNotificationRequest = UserNotificationRequest.builder()
+					.recipientUserId(Env.getLoggedUserId())
+					.contentADMessage(MSG_SKIPPED_C_ORDERLINE_IDS)
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/process/C_Order_CreatePOFromSOs.java:138]: the C_OrderLine_ID don't tell the users anything. please output the `C_Order.DocumentNo` and `C_OrderLine.Line`.
`SkippedOrderLines`
- :speaking_head: EN `Skipped sales order lines {0}` => DE `Übersprungene Auftragszeilen: {0}`
- ...and the parameters could be a comma-delimited (`', '`) list of `<C_Order.DocumentNo>-<C_OrderLine.Line>`

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregator.java [hunk 6]
```diff
@@ -96,6 +102,7 @@ protected I_C_Order createGroup(final Object vendorBPartnerValue, final I_C_Orde
 	{
 		if (CreatePOFromSOsAggregationKeyBuilder.KEY_SKIP.equals(vendorBPartnerValue))
 		{
+			skippedSalesOrderLineIds.add(OrderLineId.ofRepoId(salesOrderLine.getC_OrderLine_ID()));
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregator.java:105]: please log this to `Loggables.withLogger(logger, Level.DEBUG)` (also add a static logger)

PS: ah, no..you are already doing that in `CreatePOFromSOsAggregationKeyBuilder.java`, right?

File: backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregationKeyBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	private I_M_Product productRecord;
+	private I_C_BPartner customerBPartner;
+	private I_C_Order salesOrderRecord;
+
+	@BeforeEach
+	void init()
+	{
+		AdempiereTestHelper.get().init();
+
+		productRecord = newInstance(I_M_Product.class);
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/test/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregationKeyBuilderTest.java:58]: ok, but note for next time you can also use `de.metas.business.BusinessTestHelper` to create such test data

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregationKeyBuilder.java [hunk 8]
```diff
@@ -131,10 +131,10 @@ public String buildKey(final I_C_OrderLine salesOrderLine)
 		if (p_Vendor_ID > 0 && poPartnerRepoId != p_Vendor_ID)
 		{
 			final String msg = msgBL.getMsg(context.getCtx(),
-					MSG_VENDOR_MISMATCH,
-					new Object[] { salesOrder.getDocumentNo(), salesOrderLine.getLine(), poBPartnerRecord.getValue(), poBPartnerRecord.getName() });
+											MSG_VENDOR_MISMATCH,
+											new Object[] { salesOrder.getDocumentNo(), salesOrderLine.getLine(), poBPartnerRecord.getValue(), poBPartnerRecord.getName() });
 
-			loggable.addLog(msg);
+			Loggables.get().addLog(msg);
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregationKeyBuilder.java:137]: please also add a "normal" static logger and the call Loggables.withLogger(logger, Level.DEBUG)

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/process/C_Order_CreatePOFromSOs.java (hunk 5 above)
Comment (dm-ps) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/process/C_Order_CreatePOFromSOs.java]: done

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregator.java (hunk 6 above)
Comment (dm-ps) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOFromSOsAggregator.java]: Logged in CreatePOFromSOsAggregator#addItemToGroup. We're not logging in CreatePOFromSOsAggregationKeyBuilder.java

**Review by metas-ts — APPROVED** [review]

**Pull request merged**